cli/cmd/proxy/read/command.go

```diff
@@ -222,90 +244,144 @@ func (c *ReadCommand) initKubernetes() (err error) {
 	return nil
 }
 
-func (c *ReadCommand) outputConfig(config *EnvoyConfig) {
-	if c.flagJSON {
+func (c *ReadCommand) outputConfig(config *EnvoyConfig) error {
+	switch c.flagOutput {
+	case Table:
+		c.outputAsTables(config)
+	case JSON:
+		return c.outputAsJSON(config)
+	case Raw:
 		c.UI.Output(string(config.rawCfg))
-		return
 	}
 
+	return nil
+}
+
+func (c *ReadCommand) outputAsJSON(config *EnvoyConfig) error {
+	cfg := make(map[string]interface{})
+	if !c.filtersPassed() || c.flagClusters {
+		cfg["clusters"] = config.Clusters
+	}
+	if !c.filtersPassed() || c.flagEndpoints {
+		cfg["endpoints"] = config.Endpoints
+	}
+	if !c.filtersPassed() || c.flagListeners {
+		cfg["listeners"] = config.Listeners
+	}
+	if !c.filtersPassed() || c.flagRoutes {
+		cfg["routes"] = config.Routes
+	}
+	if !c.filtersPassed() || c.flagSecrets {
+		cfg["secrets"] = config.Secrets
+	}
+
+	out, err := json.MarshalIndent(cfg, "", "\t")
+	if err != nil {
+		return err
+	}
+
+	c.UI.Output(string(out))
+	return nil
+}
+
+func (c *ReadCommand) outputAsTables(config *EnvoyConfig) {
 	c.UI.Output(fmt.Sprintf("Envoy configuration for %s in Namespace %s:", c.flagPodName, c.flagNamespace))
+	c.outputClustersTable(config.Clusters)
+	c.outputEndpointsTable(config.Endpoints)
+	c.outputListenersTable(config.Listeners)
+	c.outputRoutesTable(config.Routes)
+	c.outputSecretsTable(config.Secrets)
+}
 
-	// Track if any filters are passed in. If not, print everything; if so, only
-	// print the filters that are passed in.
-	filtersPassed := c.flagClusters || c.flagEndpoints || c.flagListeners || c.flagRoutes || c.flagSecrets
+func (c *ReadCommand) outputClustersTable(clusters []Cluster) {
+	if c.filtersPassed() && !c.flagClusters {
+		return
+	}
 
-	if !filtersPassed || c.flagClusters {
-		c.UI.Output(fmt.Sprintf("Clusters (%d)", len(config.Clusters)), terminal.WithHeaderStyle())
-		clusters := terminal.NewTable("Name", "FQDN", "Endpoints", "Type", "Last Updated")
-		for _, cluster := range config.Clusters {
-			clusters.AddRow([]string{cluster.Name, cluster.FullyQualifiedDomainName, strings.Join(cluster.Endpoints, ", "),
-				cluster.Type, cluster.LastUpdated}, []string{})
-		}
-		c.UI.Table(clusters)
-		c.UI.Output("")
+	c.UI.Output(fmt.Sprintf("Clusters (%d)", len(clusters)), terminal.WithHeaderStyle())
+	table := terminal.NewTable("Name", "FQDN", "Endpoints", "Type", "Last Updated")
```

Comment:
I don't think this is blocking, but since we can't test the output, we should consider separating the building table logic from the output logic, so that we can write assertions against the build table code.  Basically, it would be extracting lines 302-306 into another function and writing a test for that with some mocked structs.   
Same for similar functions in this file.
If we have plans to be able to assert against terminal output, that works too.

Reply:
We do! Actually in [this PR](https://github.com/hashicorp/consul-k8s/pull/1336) modified the command structs to write to a buffer instead of stdout. In that PR, I added tests that assert the correctness of the output tables.

Reply:
![oh-gnarly](https://user-images.githubusercontent.com/2481360/178885269-e0f7daa5-816b-4f5d-8e54-442228683566.gif)